### PR TITLE
Bom

### PR DESCRIPTION
**Pull request by Chimrod**
Here is a pull request for the issue #36.

The request allow to ignore the first bytes in the files if they match a know BOM pattern (actually UTF-8 and UTF-16 are checked). Without this check, thoses bytes are considered as a part of the first row.

The check is done in a pure function `check_bom`, which return the number of bytes to ignore. If we want to going further, we could also return the associated encoding in order to inform the caller of the file encoding (not implemented here). I’ve plugged the call in the function `fill_in_buf_or_Eof` which is where the data are read and stored.

I’ve added a test which fails without the code (see the issue for that).

### CODE REVIEW

**Comment by Chimrod** [comment]
Hello, do you have any update on this pull request ?

File: src/csv.pp.ml [hunk 1]
```diff
@@ -120,12 +120,51 @@ type in_channel = {
...
+    ]
+
+(** check_bom [ic] [bom] return [None] if the stream does not start with a bom
+    and the number of bytes to ignore otherwise.
+ *)
+let check_bom
+ : in_channel -> (char array) list -> int option 
+ = fun ic bom -> 
+     let matched = 
+         List.fold_left (fun matched bom ->
```
Comment (Chris00) [inline src/csv.pp.ml:144]: Since you want to bail out as soon as you found a match, `find_map` seems more appropriate.

File: src/csv.pp.ml [hunk 2]
```diff
@@ -120,12 +120,51 @@ type in_channel = {
...
+         List.fold_left (fun matched bom -> 
+           match matched with 
+           | Some _ -> matched
+           | None -> 
+             let length = Array.length bom in
+             if (ic.in1 - ic.in0) < length then (None)
+             else
+               (* We are supposed to be in the very begining of the file, 
+                  and in0 should be equal to 0. *)
+               let idx = ref (ic.in0) in
```
Comment (Chris00) [inline src/csv.pp.ml:153]: Braces are unnecessary.

File: src/csv.pp.ml [hunk 3]
```diff
@@ -120,12 +120,51 @@ type in_channel = {
...
+               let equal = Array.for_all 
+                 (fun c -> 
+                     let res = Char.equal c (Bytes.get ic.in_buf !idx) in
```
Comment (Chris00) [inline src/csv.pp.ml:156]: Wouldn't
```ocaml
let b = Bytes.get ic.in_buf !idx in
incr idx;
Char.equal b c
```
be clearer?

**Review by Chris00 — CHANGES_REQUESTED** [review]

**Comment by Chimrod** [comment]
Hi ! Thanks for your review and feedback. I will check all the points and answer them quickly.

**Comment by SGrondin** [comment]
@Chimrod I've rebased your commits and finished the PR in #41 

Thank you for your contribution!

**Pull request closed without merging**